src/client.c

```diff
@@ -336,16 +336,23 @@ int trilogy_connect_recv(trilogy_conn_t *conn, trilogy_handshake_t *handshake_ou
 int trilogy_auth_send(trilogy_conn_t *conn, const trilogy_handshake_t *handshake)
 {
     trilogy_builder_t builder;
+    const char *auth_plugin;
 
     int rc = begin_command_phase(&builder, conn, conn->packet_parser.sequence_number);
 
     if (rc < 0) {
         return rc;
     }
 
+    if (conn->socket->opts.enable_cleartext_plugin) {
+        auth_plugin = "mysql_clear_password";
+    } else {
+        auth_plugin = handshake->auth_plugin;
+    }
```

Comment:
Just so I understand correctly, you're asking if the server will accept the cleartext plugin password without the client informing the server of the plugin? 

If so, then I believe that to be the case. According to the `libmysqlclient` [source code / documentation](https://github.com/mysql/mysql-server/blob/824e2b4064053f7daf17d7f3f84b7a3ed92e5fb4/sql-common/client.cc#L9454-L9512), we don't need to preemptively inform the server that we are using the cleartext plugin before the authentication connection. To put it in another way, if cleartext plugin is enabled, it does become the "default" that the client advertises, in a sense where the client can just send it to the server.

Reply:
What happens if we connect with `enable_cleartext_plugin :true` but then the server handshake indicates a non-cleartext plugin like caching_sha2_password or whatever? Will this code ignore what the server told us and send the cleartext password anyway? Does that then turn into an auth switch?

(I'm still learning about how these plugins work, so not sure I'm making sense.)

Reply:
You're right, the first handshake we get back from the server will indicate the server's authentication plugin preference. However, according to [the MySQL docs on "Determining Authentication Method"](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase.html#sect_protocol_connection_phase_initial_handshake_auth_method), the client does not have to use the server's authentication method. 

> When including authentication reply in the [Protocol::HandshakeResponse:](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_packets_protocol_handshake_response.html), client is not obligated to use the same authentication method that was used by the server in the [Protocol::Handshake](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_packets_protocol_handshake.html) packet.

As the code is right now, we will send the cleartext password but the server can still respond with an auth switch request, and we will do so accordingly. We think this is OK because the request from the user to use cleartext password has been fulfilled. If the server is still not happy then it's OK for us to do the auth switch.

Reply:
Do you have a good way of testing out one of the plugins that needs a cleartext password? I'm curious what plugin is advertised in the handshake for those. Like, if the server wants to use `authentication_pam` does the handshake advertise the server plugin `authentication_pam` or the client plugin `mysql_clear_password`?

Reply:
By the way, I should have started by saying thank you for working on this!

@hparker and I found a test server plugin that uses mysql_clear_password as the client plugin, and I was able to enable it in a locally with:

```
INSTALL PLUGIN cleartext_plugin_server SONAME 'auth_test_plugin.so';
CREATE USER 'cleartext_user'@'localhost' IDENTIFIED WITH cleartext_plugin_server BY 'password';
```

Then I connected via the command line client and looked at the packets:

```
mysql --protocol=tcp --ssl-mode=disable -p -u cleartext_user --enable-cleartext-plugin
```

1. Server sends handshake, with auth plugin caching_sha2_password
2. Client sends handshake response, **still with auth plugin caching_sha2_password**
3. Server sends auth switch request, now with auth plugin `mysql_clear_password`
4. Client sends auth switch response with just the cleartext password and nothing else

That's more what I was expecting. The server is telling the client to use `mysql_clear_password`. While a client can choose a different plugin than what the server advertised in the handshake, I don't know how the server handles `mysql_clear_password` in particular since it is a client-side-only plugin.

Trying to connect to this same user with Trilogy using this PR results in:

```
Trilogy.new(host: '127.0.0.1', username: 'cleartext_user', password: 'password', enable_cleartext_plugin: true)
```

1. Server sends handshake, with auth plugin caching_sha2_password
2. Client sends handshake response, switching to `mysql_clear_password`
3. Server responds with "Access denied"

I'm not totally sure why that fails (including with the fix so we send the password). But maybe we should try to behave more like `mysql` here?